webhook_to_fedora_messaging/auth.py

```diff
@@ -0,0 +1,57 @@
+from asyncio import run
+
+from fastapi import Depends, HTTPException, Security
+from fastapi.security import HTTPBasic, HTTPBasicCredentials
+from sqlalchemy import select
+from sqlalchemy.orm import selectinload
+from sqlalchemy.orm.session import Session
+from starlette.status import HTTP_401_UNAUTHORIZED, HTTP_403_FORBIDDEN
+
+from webhook_to_fedora_messaging.database import get_session
+from webhook_to_fedora_messaging.models import User
+
+
+async def user_factory(optional: bool = False, **kwargs):
+    """
+    Factory creating FastAPI dependencies for authenticating users
+    """
+    if optional:
+        kwargs["auto_error"] = False
+    security = HTTPBasic(realm="W2FM", **kwargs)
+
+    async def user_actual(
+        session: Session = Depends(get_session),  # noqa : B008
+        cred: HTTPBasicCredentials = Security(security)  # noqa : B008
+    ):
+        if not cred:
+            if not optional:
+                raise HTTPException(HTTP_403_FORBIDDEN)
+            else:
+                return None
+        username, password = cred.username, cred.password
+
+        query = select(User).filter_by(username=username).options(selectinload("*"))
+        result = await session.execute(query)
+        user_data = result.scalar_one_or_none()
+
+        if not user_data:
+            raise HTTPException(
+                HTTP_401_UNAUTHORIZED,
+                "Username not found"
+            )
+
+        # Do not worry - We will move on from basic HTTP authentication to OpenID Connect
+        if password != username:
+            raise HTTPException(
+                HTTP_403_FORBIDDEN,
+                "Invalid credentials"
+            )
+
+        return user_data
+
+    return user_actual
+
+
+get_user = run(user_factory())
```

Comment:
That does not seem to work.

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib/python3.12/multiprocessing/spawn.py", line 122, in spawn_main
    exitcode = _main(fd, parent_sentinel)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/multiprocessing/spawn.py", line 131, in _main
    prepare(preparation_data)
  File "/usr/lib/python3.12/multiprocessing/spawn.py", line 246, in prepare
    _fixup_main_from_path(data['init_main_from_path'])
  File "/usr/lib/python3.12/multiprocessing/spawn.py", line 297, in _fixup_main_from_path
    main_content = runpy.run_path(main_path,
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen runpy>", line 286, in run_path
  File "<frozen runpy>", line 98, in _run_module_code
  File "<frozen runpy>", line 88, in _run_code
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/bin/w2fm", line 3, in <module>
    from webhook_to_fedora_messaging.clim import main
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/webhook_to_fedora_messaging/clim.py", line 7, in <module>
    from webhook_to_fedora_messaging.main import start_service
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/webhook_to_fedora_messaging/main.py", line 14, in <module>
    from webhook_to_fedora_messaging.endpoints import message, service, user
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/webhook_to_fedora_messaging/endpoints/user.py", line 31, in <module>
    @router.post(
     ^^^^^^^^^^^^
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/fastapi/routing.py", line 944, in decorator
    self.add_api_route(
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/fastapi/routing.py", line 883, in add_api_route
    route = route_class(
            ^^^^^^^^^^^^
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/fastapi/routing.py", line 513, in __init__
    self.dependant = get_dependant(path=self.path_format, call=self.endpoint)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/fastapi/dependencies/utils.py", line 268, in get_dependant
    sub_dependant = get_param_sub_dependant(
                    ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/fastapi/dependencies/utils.py", line 111, in get_param_sub_dependant
    return get_sub_dependant(
           ^^^^^^^^^^^^^^^^^^
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/fastapi/dependencies/utils.py", line 147, in get_sub_dependant
    sub_dependant = get_dependant(
                    ^^^^^^^^^^^^^^
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/fastapi/dependencies/utils.py", line 250, in get_dependant
    endpoint_signature = get_typed_signature(call)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/fastapi/dependencies/utils.py", line 208, in get_typed_signature
    signature = inspect.signature(call)
                ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/inspect.py", line 3335, in signature
    return Signature.from_callable(obj, follow_wrapped=follow_wrapped,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/inspect.py", line 3075, in from_callable
    return _signature_from_callable(obj, sigcls=cls,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/inspect.py", line 2512, in _signature_from_callable
    raise TypeError('{!r} is not a callable object'.format(obj))
TypeError: <coroutine object user_factory at 0x70504c7563e0> is not a callable object
sys:1: RuntimeWarning: coroutine 'user_factory' was never awaited
```

Reply:
Strangely enough - I am not getting the error that you stated.

```
RuntimeError: asyncio.run() cannot be called from a running event loop
```

Having `asyncio.run` execute multiple times is anti-pattern but the following tries to instantiate the `user_factory()` synchronously, meaning that it is not awaited.

```
async def create_user(
    body: UserRequest,
    session: AsyncSession = Depends(get_session),  # noqa : B008
    user: User = Depends(user_factory())  # noqa : B008
):
```

And when I try to remove the optional authentication condition as a parameter for the dependency, the following removes the authentication entirely.

```
async def create_user(
    body: UserRequest,
    session: AsyncSession = Depends(get_session),  # noqa : B008
    user: User = Depends(user_factory)  # noqa : B008
):
```

Reply:
That's because the `user_factory` function must not be async. It returns `user_actual` which is async, but `user_factory` itself must not be async.

You could do it in a class-based manner if you prefer:

```py
class UserAuth:
    def __init__(self, optional: bool = False, **kwargs):
        self._optional = optional
        if self._optional:
            kwargs["auto_error"] = False
        self._security = HTTPBasic(realm="W2FM", **kwargs)

    async def __call__(
        session: Session = Depends(get_session),  # noqa : B008
        cred: HTTPBasicCredentials = Security(self._security),  # noqa : B008
    ):
        if not cred:
            if not self._optional:
                raise HTTPException(HTTP_403_FORBIDDEN)
            else:
                return None
        username, password = cred.username, cred.password

        query = select(User).filter_by(username=username)
        result = await session.execute(query)
        try:
            user_data = result.scalar_one()
        except NoResultFound:
            raise HTTPException(HTTP_401_UNAUTHORIZED, "Username not found")

        # Do not worry - We will move on from basic HTTP authentication to OpenID Connect
        if password != username:
            raise HTTPException(HTTP_403_FORBIDDEN, "Invalid credentials")

        return user_data
```

And then you depend on `UserAuth()` or `UserAuth(optional=True)`.

Reference: https://fastapi.tiangolo.com/advanced/advanced-dependencies/

Reply:
ISTG 😠 

```
[W2FM] [2024-08-08 01:11:29 +0530] [ERROR] Exception in ASGI application
Traceback (most recent call last):
  File "asyncpg/protocol/prepared_stmt.pyx", line 175, in asyncpg.protocol.protocol.PreparedStatementState._encode_bind_msg
  File "asyncpg/protocol/codecs/base.pyx", line 227, in asyncpg.protocol.protocol.Codec.encode
  File "asyncpg/protocol/codecs/base.pyx", line 129, in asyncpg.protocol.protocol.Codec.encode_scalar
  File "asyncpg/pgproto/./codecs/datetime.pyx", line 152, in asyncpg.pgproto.pgproto.timestamp_encode
TypeError: can't subtract offset-naive and offset-aware datetimes

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 538, in _prepare_and_execute
    self._rows = deque(await prepared_stmt.fetch(*parameters))
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/asyncpg/prepared_stmt.py", line 176, in fetch
    data = await self.__bind_execute(args, 0, timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/asyncpg/prepared_stmt.py", line 241, in __bind_execute
    data, status, _ = await self.__do_execute(
                      ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/asyncpg/prepared_stmt.py", line 230, in __do_execute
    return await executor(protocol)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "asyncpg/protocol/protocol.pyx", line 186, in bind_execute
  File "asyncpg/protocol/prepared_stmt.pyx", line 204, in asyncpg.protocol.protocol.PreparedStatementState._encode_bind_msg
asyncpg.exceptions.DataError: invalid input for query argument $4: datetime.datetime(2024, 8, 8, 7, 41, 29,... (can't subtract offset-naive and offset-aware datetimes)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1967, in _exec_single_context
    self.dialect.do_execute(
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/engine/default.py", line 924, in do_execute
    cursor.execute(statement, parameters)
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 572, in execute
    self._adapt_connection.await_(
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 132, in await_only
    return current.parent.switch(awaitable)  # type: ignore[no-any-return,attr-defined] # noqa: E501
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 196, in greenlet_spawn
    value = await result
            ^^^^^^^^^^^^
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 550, in _prepare_and_execute
    self._handle_exception(error)
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 501, in _handle_exception
    self._adapt_connection._handle_exception(error)
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 784, in _handle_exception
    raise translated_error from error
sqlalchemy.dialects.postgresql.asyncpg.AsyncAdapt_asyncpg_dbapi.Error: <class 'asyncpg.exceptions.DataError'>: invalid input for query argument $4: datetime.datetime(2024, 8, 8, 7, 41, 29,... (can't subtract offset-naive and offset-aware datetimes)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/uvicorn/protocols/http/httptools_impl.py", line 399, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/uvicorn/middleware/proxy_headers.py", line 70, in __call__
    return await self.app(scope, receive, send)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/fastapi/applications.py", line 1054, in __call__
    await super().__call__(scope, receive, send)
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/starlette/applications.py", line 123, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/starlette/middleware/errors.py", line 186, in __call__
    raise exc
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/starlette/middleware/errors.py", line 164, in __call__
    await self.app(scope, receive, _send)
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/starlette/middleware/exceptions.py", line 65, in __call__
    await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
    raise exc
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    await app(scope, receive, sender)
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/starlette/routing.py", line 756, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/starlette/routing.py", line 776, in app
    await route.handle(scope, receive, send)
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/starlette/routing.py", line 297, in handle
    await self.app(scope, receive, send)
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/starlette/routing.py", line 77, in app
    await wrap_app_handling_exceptions(app, request)(scope, receive, send)
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
    raise exc
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    await app(scope, receive, sender)
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/starlette/routing.py", line 72, in app
    response = await func(request)
               ^^^^^^^^^^^^^^^^^^^
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/fastapi/routing.py", line 278, in app
    raw_response = await run_endpoint_function(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/fastapi/routing.py", line 191, in run_endpoint_function
    return await dependant.call(**values)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/webhook_to_fedora_messaging/endpoints/user.py", line 51, in create_user
    await session.flush()
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/ext/asyncio/session.py", line 800, in flush
    await greenlet_spawn(self.sync_session.flush, objects=objects)
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 203, in greenlet_spawn
    result = context.switch(value)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/orm/session.py", line 4341, in flush
    self._flush(objects)
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/orm/session.py", line 4476, in _flush
    with util.safe_reraise():
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/util/langhelpers.py", line 146, in __exit__
    raise exc_value.with_traceback(exc_tb)
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/orm/session.py", line 4437, in _flush
    flush_context.execute()
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/orm/unitofwork.py", line 466, in execute
    rec.execute(self)
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/orm/unitofwork.py", line 642, in execute
    util.preloaded.orm_persistence.save_obj(
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/orm/persistence.py", line 93, in save_obj
    _emit_insert_statements(
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/orm/persistence.py", line 1233, in _emit_insert_statements
    result = connection.execute(
             ^^^^^^^^^^^^^^^^^^^
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1418, in execute
    return meth(
           ^^^^^
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/sql/elements.py", line 515, in _execute_on_connection
    return connection._execute_clauseelement(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1640, in _execute_clauseelement
    ret = self._execute_context(
          ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1846, in _execute_context
    return self._exec_single_context(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1986, in _exec_single_context
    self._handle_dbapi_exception(
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 2353, in _handle_dbapi_exception
    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1967, in _exec_single_context
    self.dialect.do_execute(
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/engine/default.py", line 924, in do_execute
    cursor.execute(statement, parameters)
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 572, in execute
    self._adapt_connection.await_(
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 132, in await_only
    return current.parent.switch(awaitable)  # type: ignore[no-any-return,attr-defined] # noqa: E501
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 196, in greenlet_spawn
    value = await result
            ^^^^^^^^^^^^
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 550, in _prepare_and_execute
    self._handle_exception(error)
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 501, in _handle_exception
    self._adapt_connection._handle_exception(error)
  File "/home/archdesk/Projects/webhook-to-fedora-messaging/venv/lib/python3.12/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 784, in _handle_exception
    raise translated_error from error
sqlalchemy.exc.DBAPIError: (sqlalchemy.dialects.postgresql.asyncpg.Error) <class 'asyncpg.exceptions.DataError'>: invalid input for query argument $4: datetime.datetime(2024, 8, 8, 7, 41, 29,... (can't subtract offset-naive and offset-aware datetimes)
[SQL: INSERT INTO users (username, is_admin, uuid, creation_date) VALUES ($1::VARCHAR, $2::BOOLEAN, $3::VARCHAR, $4::TIMESTAMP WITHOUT TIME ZONE) RETURNING users.id]
[parameters: ('string', False, 'ae1acf06', datetime.datetime(2024, 8, 8, 7, 41, 29, 439303, tzinfo=datetime.timezone.utc))]
(Background on this error at: https://sqlalche.me/e/20/dbapi)
```

Reply:
Did you rebuild the DB ? What does this say:
```
$ psql w2fm
w2fm=> \d users
```
The type of `creation_date` should be `timestamp with time zone`.

Reply:
GAH it's `TIMESTAMP(timezone=True)` :-/

Reply:
it was all in SQLAlchemy's docs but I didn't read it properly

Reply:
I nuked the database and recreated it.

Following the most recent comment, I get the following:

![image](https://github.com/user-attachments/assets/a009260a-0621-4a0d-8eec-12330ed8ec09)

I use `tmux` and hence the screenshot.

Reply:
And it worked.

```
{
  "data": {
    "uuid": "79a01f9b",
    "username": "string",
    "is_admin": false,
    "creation_date": "2024-08-08T08:21:30.207222Z"
  }
}
```

```
w2fm=# \d+ users
```

```
                                                                   Table "public.users"
    Column     |           Type           | Collation | Nullable |              Default              | Storage  | Compression | Stats target | Description
---------------+--------------------------+-----------+----------+-----------------------------------+----------+-------------+--------------+-------------
 id            | integer                  |           | not null | nextval('users_id_seq'::regclass) | plain    |             |              |
 username      | text                     |           | not null |                                   | extended |             |              |
 is_admin      | boolean                  |           | not null |                                   | plain    |             |              |
 uuid          | text                     |           | not null |                                   | extended |             |              |
 creation_date | timestamp with time zone |           | not null |                                   | plain    |             |              |
Indexes:
    "pk_users" PRIMARY KEY, btree (id)
    "uq_users_username" UNIQUE CONSTRAINT, btree (username)
    "uq_users_uuid" UNIQUE CONSTRAINT, btree (uuid)
Referenced by:
    TABLE "services" CONSTRAINT "fk_services_user_id_users" FOREIGN KEY (user_id) REFERENCES users(id) ON DELETE CASCADE
Access method: heap
```

```
w2fm=# select * from public.users;
```

```
 id | username | is_admin |   uuid   |         creation_date
----+----------+----------+----------+-------------------------------
  1 | testuser | f        | deadbeef | 2024-08-02 10:35:27.60185+00
  2 | string   | f        | 79a01f9b | 2024-08-08 08:21:30.207222+00
(2 rows)
```